src/keyfile.c

```diff
@@ -1313,11 +1316,47 @@ void configuration_apply_settings(void)
 }
 
 
+static gboolean save_configuration_cb(gpointer data)
+{
+	configuration_save();
+	if (app->project != NULL)
+	{
+		project_write_config();
+	}
+	document_list_update_idle_func_id = 0;
+	return G_SOURCE_REMOVE;
+}
+
+
+static void document_list_changed_cb(GObject *obj, GeanyDocument *doc, gpointer data)
+{
+	g_return_if_fail(doc != NULL && doc->is_valid);
+
+	/* save configuration, especially session file list, but only if we are not just starting
+	 * and not about to quit */
+	if (file_prefs.save_config_on_file_change &&
+		main_status.main_window_realized &&
+		!main_status.opening_session_files &&
+		!main_status.quitting)
+	{
+		if (document_list_update_idle_func_id == 0)
+		{
+			document_list_update_idle_func_id = g_idle_add(save_configuration_cb, NULL);
+		}
+	}
+}
+
+
 void configuration_init(void)
 {
 	keyfile_groups = g_ptr_array_new();
 	pref_groups = g_ptr_array_new();
 	init_pref_groups();
+
+	g_signal_connect(geany_object, "document-new", G_CALLBACK(document_list_changed_cb), NULL);
+	g_signal_connect(geany_object, "document-open", G_CALLBACK(document_list_changed_cb), NULL);
+	g_signal_connect(geany_object, "document-save", G_CALLBACK(document_list_changed_cb), NULL);
```

Comment:
It would be nice to avoid saving the conf when saving a file in the common case, e.g. when just updating the file's content.  AFAIK, we don't affect the session then, do we?  My point is that I'm usually saving my files very often, and there's no use of saving the config then, only if I saved as another name or such, is there?
I agree it's probably cheap to save it anyway (although it might matter a little on SSDs?), but if it's of no use at all I'd rather see it avoided.

Not sure how to implement that though… the only thing I can think of is to check in `:document-before-save` that `doc->real_path == NULL` and then set a flag to handle that one in `:document-save`, for when saving is successful?  Or just don't bother to check if save worked, and just do it in `:document-before-save` with the check, as the likely case is a successful save anyway.

---
Untested possible implementations:

The simple solution:
```diff
+static void document_before_save_cb(GObject *obj, GeanyDocument *doc, gpointer data)
+{
+	if (! doc->real_path)
+		document_list_changed_cb(obj, doc, data);
+}
+
/* … */
-	g_signal_connect(geany_object, "document-save", G_CALLBACK(document_list_changed_cb), NULL);
+	g_signal_connect(geany_object, "document-before-save", G_CALLBACK(document_before_save_cb), NULL);
/* … */
```

The convoluted solution:

```diff
+static void document_before_save_cb(GObject *obj, GeanyDocument *doc, gpointer data)
+{
+	gboolean *saved_as = data;
+	*saved_as = doc->real_path == NULL;
+}
+
+static void document_save_cb(GObject *obj, GeanyDocument *doc, gpointer data)
+{
+	gboolean *saved_as = data;
+	if (*saved_as)
+		document_list_changed_cb(obj, doc, NULL);
+}
+
/* … */
+	gboolean *saved_as_p = g_malloc(sizeof *saved_as_p);
+	g_signal_connect_data(geany_object, "document-before-save", G_CALLBACK(document_before_save_cb), saved_as_p, (GClosureNotify) g_free, 0);
-	g_signal_connect(geany_object, "document-save", G_CALLBACK(document_list_changed_cb), NULL);
+	g_signal_connect(geany_object, "document-save", G_CALLBACK(document_list_changed_cb), saved_as_p);
/* … */
```

Reply:
Well, the session includes cursor position so its probably worth saving.

Reply:
Why is document-new connected to? If there's no file yet it won't affect the session.

Reply:
> It would be nice to avoid saving the conf when saving a file in the common case, e.g. when just updating the file's content. AFAIK, we don't affect the session then, do we? My point is that I'm 

Hmm, I'm not sure if checking `doc->real_path` for `NULL` is sufficient, there might be more cases when we want to trigger the config save. Also, the `doc->real_path` check would have to differentiate cases like renaming and closing documents.
The config save process is pretty lightweight and, as said initially, very fast.
Users trying to reduce disk IO still can disable this feature.

Can we agree to implement more sophisticated checks independent of this PR afterwards if someone wants to?

> Why is document-new connected to? If there's no file yet it won't affect the session.

Removed in https://github.com/geany/geany/pull/2114/commits/d6ebff6c4f6c04786e5448735ac59462e2f7619a. Probably not necessary, I just added it for completeness.